FAB-17839 Ch.Part.API: List channels REST handler

tock-ibm opened the pull request:
Develop the following handler functionality.

List all channels:
`GET /participation/v1/channels`.

List a single channel:
`GET /participation/v1/channels/my-channel`.

Define the interface the Registrar will implement,
place skeleton methods in it.

Signed-off-by: Yoav Tock <tock@il.ibm.com>
Change-Id: I4ce65b406ccbbbe8e1dbd4a95680c3561f5fc7c9

<!--- DELETE MARKDOWN COMMENTS BEFORE SUBMITTING PULL REQUEST. -->

<!--- Provide a descriptive summary of your changes in the Title above. -->

#### Type of change

<!--- What type of change? Pick one option and delete the others. -->

- New feature

#### Description

<!--- Describe your changes in detail, including motivation. -->

#### Additional details

<!--- Additional implementation details or comments to reviewers. -->
<!--- Summarize how the pull request was tested (if not obvious from commit). -->

#### Related issues
Story: https://jira.hyperledger.org/browse/FAB-15710
Epic: https://jira.hyperledger.org/browse/FAB-17712

<!--- Include a link to any associated issues, e.g. Jira issue or approved rfc. -->

<!---
#### Release Note
If change impacts current users, uncomment Release Note heading and provide
release note text.
Also, copy release note text into the release specific /release_notes file.
-->

<!--
Checklist (DELETE AFTER READING):

- `Signed-off-by` added to commits (required for DCO check to pass)
- Tests have been added/updated (required for bug fixes and features)
- Unit and/or integration tests pass locally
- Run linters and checks locally using 'make checks'
- If change requires documentation updates, make updates in pull request,
  or open a separate issue and provide link
- Squash commits into a single commit, unless a stack of commits is
  intentional to assist reviewers or to preserve review comments.
- For additional contribution guidelines see the project's CONTRIBUTING.md file
-->

tock-ibm committed:
FAB-17839 Ch.Part.API: List REST handler

Develop the following handler functionality.

List all channels:
`GET /participation/v1/channels`.

List a single handler:
`GET /participation/v1/channels/my-channel`.

Define the interface the Registrar will implement,
place skeleton methods in it.

Signed-off-by: Yoav Tock <tock@il.ibm.com>
Change-Id: I4ce65b406ccbbbe8e1dbd4a95680c3561f5fc7c9

tock-ibm committed:
fix gofmt

Signed-off-by: Yoav Tock <tock@il.ibm.com>
Change-Id: I4836ff6dc3101c1ceaeb98db8234ee732f511717

yacovm commented on orderer/common/multichannel/registrar.go:
```diff
@@ -370,3 +371,13 @@ func (r *Registrar) NewChannelConfig(envConfigUpdate *cb.Envelope) (channelconfi
 	return channelconfig.NewBundle(channelID, config, r.bccsp)
 }
+
+func (r *Registrar) ListAllChannels() ([]channelparticipation.ChannelInfoShort, string) {
```
I'm not sure the registrar should return a slice of this type. This creates an unnecessary dependency on this package. Perhaps define a type in a leaf package and then have the registrar return a slice of that type?

tock-ibm replied:
I was thinking the same way and originally had the `ListAllChannels` return  `([]string, string)`.... But then, when I crafted the 

`ListChannel(channelID string) (*channelparticipation.ChannelInfoFull, error)` 

I realized these two packages are going to be linked anyway... so I gave up on that direction. 
The way it is now we don't have to copy strings from one struct to another...

Putting these types in a leaf package to break the cycle will help.

yacovm commented on orderer/common/multichannel/registrar.go:
```diff
@@ -370,3 +371,13 @@ func (r *Registrar) NewChannelConfig(envConfigUpdate *cb.Envelope) (channelconfi
+	return make([]channelparticipation.ChannelInfoShort,0), ""
+}
+
+func (r *Registrar) ListChannel(channelID string) (*channelparticipation.ChannelInfoFull, error) {
```
Perhaps indicate the system channel string into the struct returned from this API as a boolean?

tock-ibm replied:
We want to quickly identify whether we have one or not w/o scanning the entire list... this goes to the client so we want this efficient and clear. Note that the system channel may change from "" to "i-am-here" and back to "" as we can create and delete it with this API.

yacovm replied:
Under what circumstance do we need to identify if we have a system channel or not, and not also anyway return all channels? 

The way this is used in the existing code in this PR is that you return all the channels anyway, so you are going to see who is the system channel when you iterate over it anyway, no?

tock-ibm replied:
Yes, but this structure is intended mainly for the client, who receives JSON.
The client may want to identify the system channel immediately, for example, when it needs to know if it is at all possible to create channels or only list them. If the system channel exists, the API can only serve GET and DELETE of the system channel. This way, the client can find that out without iterating.

tock-ibm replied:
Moreover, this representation is shorter. 

```
{
  "size":3,
  "systemChannel":"a",
  "channels":[
     {"name":"a","url":"/api/channels/a"},
     {"name":"b","url":"/api/channels/b"},
     {"name":"c","url":"/api/channels/c"},
]}
```

is shorter than 


```
{
  "size":2,
  "channels":[
     {"name":"a","url":"/api/channels/a","systemChannel":"true"}, 
     {"name":"b","url":"/api/channels/b","systemChannel":"false"},
     {"name":"c","url":"/api/channels/c","systemChannel":"false"},
]}
```
even for a short list.

tock-ibm replied:
The other alternative I see is:

```
{
  "systemChannel":{"name":"a","url":"/api/channels/a"},
  "channels":[
     {"name":"b","url":"/api/channels/b"},
     {"name":"c","url":"/api/channels/c"},
  "size":2,
]}
```
i.e. that the `channels` array carry only application channels.

yacovm replied:
I like the last one the most, modulo the size which is redundant.

C0rWin replied:
There is another option to return a map:

```
{
  "systemChannel": "a",
  "channels": {
      "a": "/api/channels/a",
      "b": "/api/channels/b",
      "c": "/api/channels/c",
   }
}
```

tock-ibm replied:
OK Lets's go with 
```
{
  "systemChannel":{"name":"a","url":"/api/channels/a"},
  "channels":[
     {"name":"b","url":"/api/channels/b"},
     {"name":"c","url":"/api/channels/c"},
]}
```
I want to keep the ChannelInfoShort struct just in case we'll need to add another field one day.

C0rWin commented on orderer/common/channelparticipation/restapi.go:
```diff
@@ -51,7 +58,43 @@ func (h *HTTPHandler) ServeHTTP(resp http.ResponseWriter, req *http.Request) {
+			return
+		}
+
+		channelID, all, err := detectResource(req.URL.Path)
```
this is not the way you should implement this logic, instead of manually extracting parameters you should template REST endpoint, for example:

```
  var channelsEndpoint = /participation/v1/channels/{channelID}
```

then you can use

```
   vars := mux.Vars(req)
   channelID := vars["channelID"]
```

C0rWin replied:
just realized we are not using `gorilla/mux` package to route REST requests.

tock-ibm replied:
True! We have github.com/gorilla already vendored in Fabric... I wan't aware of that. I'll use that.

C0rWin commented on orderer/common/channelparticipation/responses.go:
```diff
@@ -0,0 +1,34 @@
+
+// ListAllChannels carries the response to a List of all channels.
+type ListAllChannels struct {
+	Size          int                `json:"size"`          // The number of channels
```
Why do you need this? Isn't it just length of the `Channels` slice?

tock-ibm replied:
It is, but it goes to the client via JSON, so maybe it is useful there?

tock-ibm replied:
See comment to @yacovm below...

tock-ibm replied:
Out it goes.

C0rWin commented on orderer/common/channelparticipation/restapi.go:
```diff
@@ -9,21 +9,28 @@ package channelparticipation
 type ChannelManagement interface {
+	// ListAll returns the names of all channels (including the system channel), and the name of the system channel
+	// (empty if does not exist). The URL field is empty, and is to be completed by the caller.
+	ListAllChannels() ([]ChannelInfoShort, string)
```
IMO, instead of returning a touple it's better to wrap it into struct.

C0rWin replied:
Also I think that better name just to call it `Channels`

tock-ibm replied:
OK

C0rWin commented on orderer/common/channelparticipation/restapi.go:
```diff
@@ -51,7 +58,43 @@ func (h *HTTPHandler) ServeHTTP(resp http.ResponseWriter, req *http.Request) {
+			channels.Channels, channels.SystemChannel = h.registrar.ListAllChannels()
+			channels.Size = len(channels.Channels)
+			for i, info := range channels.Channels {
+				channels.Channels[i].URL = URLBaseV1Channels + `/` + info.Name
```
```
path.Join(URLBaseV1Channels, info.Name)
```

tock-ibm replied:
OK

C0rWin commented on orderer/common/channelparticipation/restapi.go:
```diff
@@ -9,21 +9,28 @@ package channelparticipation
+
+	// ListChannel provides extended status information about a channel.
+	// The URL field is empty, and is to be completed by the caller.
+	ListChannel(channelID string) (*ChannelInfoFull, error)
```
`ChannelInfo`?

tock-ibm replied:
OK

C0rWin commented on orderer/common/channelparticipation/restapi.go:
```diff
@@ -51,7 +58,43 @@ func (h *HTTPHandler) ServeHTTP(resp http.ResponseWriter, req *http.Request) {
+			h.sendResponseJsonError(resp, http.StatusBadRequest, errors.Wrap(err, "invalid channel ID"))
+			return
+		}
+		infoFull, err := h.registrar.ListChannel(channelID)
```
I think there is a way to register different path patterns to distinguish the desired function call, that is to separate logic of list all channels versus retrieving specific channel information.

tock-ibm replied:
True! using `github.com/gorilla/mux`, I'll use that.

tock-ibm committed:
Review comments: split types to leaf package

Split the types used by packages multichannel & channelparticipation
to a leaf package, in order to minimize dependencies.

Signed-off-by: Yoav Tock <tock@il.ibm.com>
Change-Id: I0b9ea17fd85f1f5bf1aba3ea8be80fc7b19e190f

tock-ibm committed:
Review comments: improve ChannelManagement interface

Streamline the ChannelManagement interface return types and type names.

Signed-off-by: Yoav Tock <tock@il.ibm.com>
Change-Id: I67e105f9e7e315c1ce6f5b16898de4aa94e609eb

tock-ibm committed:
Review comments: use gorilla/mux to route requests

Instead of parsing and matching the path, use
gorilla/mux to route requests and identify the channel-id.

Signed-off-by: Yoav Tock <tock@il.ibm.com>
Change-Id: I6739fe027fab523a0a26a25838a36e42cc7f1469

C0rWin commented on orderer/common/channelparticipation/restapi.go:
```diff
@@ -50,23 +78,122 @@ func (h *HTTPHandler) ServeHTTP(resp http.ResponseWriter, req *http.Request) {
+	}
+}
+
+func negotiateContentType(req *http.Request) (string, error) {
```
instead what you can do is to define headers you would accept while initializing gorilla/mux router, e.g.

```
    router.Headers("Content-Type", "application/json")
```

C0rWin replied:
Or there is an option to handle regexps as well


```
    router.HeadersRegexp("Content-Type", "application/(text|json)")
```

tock-ibm replied:
Good observation. I will use it to ensure that the POST arrives with Content-type: application/json. (GET and DELETE have empty bodies, so there is no need to check.)

The Headers() matcher is good for filtering out requests that come in with bad Content-Type, for example, but here I am dealing with the Accept header. It needs special treatment that is not easy with regexp, and in particular it requires a response with http.StatusNotAcceptable and an error message in the body that describes what the server can deliver. In other cases it requires selecting the Mime type according to the q value of preferred types and so on. 

See also: https://github.com/gorilla/mux/issues/18

C0rWin commented on orderer/common/channelparticipation/restapi.go:
```diff
@@ -50,23 +78,122 @@ func (h *HTTPHandler) ServeHTTP(resp http.ResponseWriter, req *http.Request) {
+
+// List all channels
+func (h *HTTPHandler) serveListAll(resp http.ResponseWriter, req *http.Request) {
+	_, err := negotiateContentType(req) // Only application/json for now
```
routers allows to define matching headers hence you can just define content type there once avoiding repeating this functionality.

tock-ibm replied:
See above, content negotiation with Accept is not really supported by the mux Headers() matcher.

C0rWin commented on orderer/common/channelparticipation/restapi.go:
```diff
@@ -50,23 +78,122 @@ func (h *HTTPHandler) ServeHTTP(resp http.ResponseWriter, req *http.Request) {
-		h.sendResponseJsonError(resp, http.StatusBadRequest, err)
+	channelID, ok := mux.Vars(req)[channelIDKey]
+	if !ok {
+		h.sendResponseJsonError(resp, http.StatusInternalServerError, errors.Wrap(err, "missing channel ID"))
```
should be ` http.StatusBadRequest` cause `http.StatusInternalServerError` means fatal error, while here is just an empty or missing channelID

tock-ibm replied:
Actually this has nothing to do with the request, it is an indication of bug - that the key used in the pattern and Vars map (const channelIDKey) is inconsistent in registration & retrieval. That would be a repeatable error internal to the server that the client cannot correct by changing the request.

C0rWin replied:
I do not understand how missing channel id in the request path is internal server error. Can you elaborate?

tock-ibm committed:
Review comments: use gorilla/mux to match headers

Signed-off-by: Yoav Tock <tock@il.ibm.com>
Change-Id: I179680a0556fef7110e69c0612687757844466e0

yacovm commented on orderer/common/types/channelinfo.go:
```diff
@@ -0,0 +1,38 @@
+	ClusterRelation string `json:"clusterRelation"` // Whether it is a “member” or ”follower”, case insensitive
+	Status          string `json:"status"`          // ”onboarding” or ”active”, case insensitive
+	Height          uint64 `json:"height"`          // Current block height
+	BlockHash       []byte `json:"blockHash"`       // Last block hash
```
What are clients expected to do with the block hash?

C0rWin replied:
For example match it with other OSNs response?

yacovm replied:
What do you expect to find, a fork in the Blockchain? 

And besides, the intended users of this interface is local administrator, you only have access to your own nodes anyway, so if a malicious leader is going to fork you, it is going to fork all your organizations nodes now that it knows you can compare hashes.

yacovm replied:
By the way let's encode this in hexa or base64 otherwise the terminal might puke funny characters if it's a byte slice.

tock-ibm replied:
It is marshaled into a JSON object as base64 by default. 
But I do have to implement a Stringer to make it log friendly, that is correct.

yacovm replied:
OK but I'm still struggling to understand why you need the hash. It only makes the output more verbose, and I doubt you can do anything useful with this hash. 

Generally, I am reluctant adding information or functionality that I am not sure how they are going to be used.

tock-ibm replied:
Adding the block-hash was a review comment I got on the [design document](https://docs.google.com/document/d/1kYoicS_Z59gNvmuuB7VFZ6WsOL31C34dWMeOFQiIygA/edit?usp=sharing), so I added it. I don't mind taking it off.

yacovm commented on orderer/common/types/channelinfo.go:
```diff
@@ -0,0 +1,38 @@
+// ChannelList carries the response to an HTTP request to List all the channels.
+// This is marshaled into the body of the HTTP response.
+type ChannelList struct {
+	Size          int                `json:"size"`          // The number of channels
```
I think having a size here is redundant as you can count the number of channels

tock-ibm replied:
I agree it is redundant from a program point of view. What I was thinking is that it would be useful for a human user that looks at the JSON object in an HTTP response, say in a browser, or using curl, to immediately tell how many channels there are. There could be many, and counting the stuff in field `Channels` could be confusing.

If you strongly object I will get rid of it.

yacovm replied:
No, you have a point.

yacovm commented on orderer/common/channelparticipation/restapi.go:
```diff
@@ -9,21 +9,35 @@ package channelparticipation
 type ChannelManagement interface {
+	// ListAll returns the names of all channels (including the system channel), and the name of the system channel
+	// (empty if does not exist). The URL field is empty, and is to be completed by the caller.
+	ListAllChannels() *types.ChannelList
```
The more idiomatic name is just **Channels()**. 
I also don't understand why the return type is a pointer type. 
Under what circumstance do we return nil? Why not return a non pointer type?

tock-ibm replied:
The pointer type is because the struct can be big, having all those channel names in there, so this way a copy is avoided.

The names corresponds to the API:

List *          -> ListAllChannels()
List x          -> ListChannel(x)
Join x         -> JoinChannel(x)
Remove x  -> RemoveChannel(x)

That is the plan, at least.

yacovm replied:
But the Channels slice inside the struct, is only initialized once, and its elements are not going to be copied even if you use a non pointer type. 

Look at this [go playground sample](https://play.golang.org/p/Cn_BPDlIJm2).
Copying a slice is cheap because you only copy 2 pointers and not the elements inside

tock-ibm replied:
You are correct. I'll change that.

yacovm commented on orderer/common/channelparticipation/restapi.go:
```diff
@@ -9,21 +9,35 @@ package channelparticipation
+
+	// ListChannel provides extended status information about a channel.
+	// The URL field is empty, and is to be completed by the caller.
+	ListChannel(channelID string) (*types.ChannelInfo, error)
```
why not non pointer type?

tock-ibm replied:
This is borderline, I agree. A non-pointer would work, since it is a "smallish" struct. However, i do return nil when the channel is not found.

yacovm replied:
if you return an error in this case then it doesn't matter if you return nil if it's not found, or a zero value

tock-ibm replied:
non- pointer type it is, then.

tock-ibm committed:
Review comments: streamline types and return values

Change return values to non-pointer type.

Streamline ChannelList:
- Drop Size,
- Change Channels slice to carry app-channels only,
- Change SystemChannel to ChannelInfoShort.

Drop BlockHash from ChannelInfo.

Signed-off-by: Yoav Tock <tock@il.ibm.com>
Change-Id: I24c6ecccb58d95c307f03332ef0a4d2c06a063aa

C0rWin commented on orderer/common/channelparticipation/restapi.go:
```diff
@@ -9,21 +9,36 @@ package channelparticipation
+	// ListAllChannels returns a slice of ChannelInfoShort containing all application channels (excluding the system
+	// channel), and ChannelInfoShort of the system channel (nil if does not exist).
+	// The URL fields are empty, and are to be completed by the caller.
+	ListAllChannels() types.ChannelList
```
I thought we agreed to rename APIs here?

tock-ibm replied:
Actually we didn't, we just agreed on the return types, see here:
https://github.com/hyperledger/fabric/pull/1232#discussion_r423875968

C0rWin replied:
I recall you have confirmed at least twice:

https://github.com/hyperledger/fabric/pull/1232#discussion_r422969830
https://github.com/hyperledger/fabric/pull/1232#discussion_r422969632

tock-ibm replied:
I change ChannelInfoLong to ChannelInfo, and used the tuple suggestion as well, but not the name of the method...

tock-ibm replied:
done

tock-ibm committed:
Review comments: rename methods

Signed-off-by: Yoav Tock <tock@il.ibm.com>
Change-Id: Ib5b88ab5bc63e4c3ed3ef1cd44c6a17e755b8b91
